"dgram" (UDP) module

pfalcon opened the pull request:
This implements basic subset of "dgram" module, per https://github.com/01org/zephyr.js/issues/386. It support both IPv4 and IPv6, and implements enough functionality for a UDP echo server (examples included).

This is now ready for review, and some points to pay to:

~~1. Some things aren't yet implemented, like buffer offset/size in .send(), args to send callback, etc.~~
2. JrS API is verbose, that's why zjs_util.c exists, but I find even that verbose too. Repetitive boilerplate code just calls to be hidden behind macros, and that's what I did. Let me know if you like, and they can be made ZJS_GET_STR(), etc., and moved to zjs_util.h . Otherwise, if you don't like it, tell me what I should do.
3. Error handling in C is also verbose and cumbersome. Why the CHECK() macro. It's not supposed to be globalized, but otherwise IMHO a useful per-module pattern.
~~4. There're still printf debugging around, it's of course supposed to be removed in the final version.~~

pfalcon commented:
This is now in much better shape: updated to use the latest goodies from net branch (and thus requires it), compiles w/o warnings, all args to .send() implemented. Perhaps the only thing I'd have in my queue is actually implement args to send's callback, and this will really require a review.

pfalcon commented:
Good news - this now can be built against Zephyr master. I've implemented the argument to .send() callback. On my todo list: 1) documentation; 2) remove debug statements; 3) squash.

pfalcon commented:
Ok, so all the points above are done, given that there were no review comments, I squashed everything to a reasonable number of commits. @grgustaf : "WIP" label can be removed and this is ready for review.

poussa commented on docs/dgram.md:
```diff
@@ -0,0 +1,98 @@
+are allowed. An example of IPv4 address: `'192.0.2.1'`, IPV6: `'2001:db8::1'`.
+
+### BLE.send
+
```
BLE ?

pfalcon replied:
Sorry, used BLE module docs as a template and overlooked some references, let me fix all of that.

pfalcon replied:
fixed

poussa commented on docs/dgram.md:
```diff
@@ -0,0 +1,98 @@
+on success.
+
+### BLE.close
+
```
ditto

pfalcon replied:
fixed

poussa commented on docs/dgram.md:
```diff
@@ -0,0 +1,98 @@
+### DgramSocket.on
+
+`void on(string event, RecvCallback callback);`
+
```
Would it be possible to support the error event also since I anticipate it will be very useful given the state of the Zephyr net stack and this new API.

pfalcon replied:
Regarding state of Z net stack: the biggest issue with it (now that it basically works) exactly that many error conditions aren't detected. Instead, it can just hang (like was due to lack of net buffers, but that's now fixed in Z version to which you guys updated Z.js recently, hurrah! I updated my code for it too), or just work in unexpected ways.

Beyond that, UDP is a simple thing and you can get only so much "errors" with it. Where you can get it, e.g. with send, I implemented error callback (which gets passed to the method). So, testing is the biggest issue here (plus some interactions which aren't obvious from Node.js docs, as we're really entering "callback hell" area with various callbacks in different places).

Anyway, if you suggest it to be done, I can go for that, test by injecting a fault by patching Z (that's what I did for send error callback), and report it as done ;-).

poussa replied:
Noted. At this point, it is more important to get this merged in so folks can start testing and using it. Then we can judge if we need better error reporting or not.

pfalcon replied:
Well, @grgustaf mentioned that you guys are busy this week preparing to OpenIoT presentation, and in few windows I have while preparing for my travel too, I'm happy to progress towards supporting "error" callback, for example I just pushed commit to make it possible to set it (but it's not called yet).

grgustaf commented on src/zjs_dgram.c:
```diff
@@ -0,0 +1,334 @@
+            args[0] = zjs_standard_error(NetworkError, errbuf);
+            // We need error object, not error value (JrS doesn't allow to
+            // pass the latter as a func argument).
+            jerry_value_clear_error_flag(&args[0]);
```
Interesting, with this comment and this line I think you understand something important that I don't understand. :) Please chat with me about this in IRC at some point.

grgustaf commented on src/zjs_dgram.c:
```diff
@@ -0,0 +1,334 @@
+        }
+
+        zjs_callback_id id = zjs_add_callback_once((jerry_value_t)user_data,
+                                      ZJS_UNDEFINED,
```
indentation - normally we line these up after the ( above.

grgustaf replied:
indentation - normally we line these up after the ( above.

pfalcon replied:
fixed

grgustaf commented on src/zjs_dgram.c:
```diff
@@ -0,0 +1,334 @@
+{
+    int ret;
+
+    if (argc < 5 || argc > 6)
```
Again we normally only check for the right minimum, and the right types - pending discussion.

Probably should check for argv[0] being object, 1-2 being numbers, etc. I guess zjs_buffer_find on the wrong type would give you NULL like you're checking for, maybe that can be skipped.

pfalcon replied:
Continuing the discussion from the previous case, as you can see, I check something, but not all. IMHO, checking number of args is a little bit more important than checking types, regarding the latter, if users pass wrong arg types, they know that they want trouble ;-).

Well, I'm joking of course, not checking types can be dangerous too, but checking them is quite verbose in Z.js :-I. I'll add them.

pfalcon replied:
> Probably should check for argv[0] being object, 1-2 being numbers, etc.

Done. Also added checking of network addr params.

grgustaf commented on src/zjs_dgram.c:
```diff
@@ -0,0 +1,334 @@
+                                          const jerry_length_t argc)
+{
+    GET_HANDLE(dgram_handle_t, handle);
+    zjs_dgram_free_cb((uintptr_t)handle);
```
I think we maybe need to set handle to NULL In the object so we don't double free? Or I guess while we're doing that, we set the free cb to NULL so it won't be called again when the socket goes out of scope.

pfalcon replied:
Yes, makes sense. I wish I knew for sure when it's safe to reset native handle though. Let me add it after the line you commented on (i.e. at the very end of close() method).

pfalcon replied:
Btw, sorry, but I don't see any other module doing that (resetting native handle explicitly). Let's just assume that they don't have explicit "close" methods and wait to be garbage-collected...

pfalcon replied:
I lied, zjs_aio.c is the only module which does that, ok.

grgustaf commented on src/zjs_dgram.c:
```diff
@@ -0,0 +1,334 @@
+        { zjs_dgram_sock_close, "close" },
+        { NULL, NULL }
+    };
+    zjs_dgram_socket_prototype = jerry_create_object();
```
Should add a cleanup function below to free this global object.

pfalcon replied:
Will do if you guys think it's important thing to do for embedded implementation.

grgustaf replied:
In our "ashell" code, we allow you to run JS files dynamically loaded from the filesystem and such. That has to clean up the whole JS state before running the next script, so it's important for that case. You do bring up a good point that for the final production image this would be wasted space, so maybe we should have a way to build that would drop all the cleanup functions. But they're probably a few bytes each at this point.

pfalcon replied:
I see, thanks for the hint. ashell is still on my todo to look at (I'm just afraid it doesn't work on anything but arduino_101, and I'll recurse into fixing that ;-) ).

pfalcon replied:
Fixed.

grgustaf commented on src/zjs_dgram.h:
```diff
@@ -0,0 +1,10 @@
+
+#include "jerry-api.h"
+
+jerry_value_t zjs_dgram_init();
```
We're trying to document new APIs (and slowly update old ones) w/ JSdoc comments here, so follow the standard comments for init/cleanup from e.g. zjs_buffer.h. Not that we actually generate these docs yet. :)

pfalcon replied:
Fixed

grgustaf commented on src/zjs_modules.c:
```diff
@@ -92,6 +93,9 @@ module_t zjs_modules_array[] = {
 #endif
 #endif // ZJS_LINUX_BUILD
+#ifdef BUILD_MODULE_DGRAM
+    { "dgram", zjs_dgram_init },
```
The cleanup will need to go in here, too.

pfalcon replied:
Fixed

grgustaf commented:
And yes, now seeing your initial comments... I do like your macros. We have had such things come and go and just haven't settled on something universal. Your GET_STR is great for the case where you have a fixed buffer where sizeof works. We should just document that carefully as it moves into zjs_util.h, as I tried to do when I boiled all our existing calls down to zjs_copy_jstring and zjs_alloc_from_jstring. That can be a follow-up patch.

I'll want to try out CHECK() myself a little to get used to it, but seems like a good one too. Another place I'd like to help with macros is the argument checking that we do so much of.

pfalcon commented on docs/dgram.md:
```diff
@@ -0,0 +1,98 @@
+
+Introduction
+------------
+The `dgram` API is based on the subset of the
```
fixed

pfalcon commented on docs/dgram.md:
```diff
@@ -0,0 +1,98 @@
+------------
+The `dgram` API is based on the subset of the
+[corresponding module](https://nodejs.org/api/dgram.html) in Node.js.
+It allows to send and receive UDP datagrams.
```
fixed

pfalcon commented on docs/dgram.md:
```diff
@@ -0,0 +1,98 @@
+entire buffer, using values `0` and `buf.length` respectively. See
+`bind()` method description for the format of `ip_addr`. An optional
+callback may be provided, which will be called with the result of send
+operation: either NetworkError object in case of error, or `undefined`
```
fixed

pfalcon commented on samples/UDPEchoServ4.js:
```diff
@@ -0,0 +1,14 @@
+var PORT = 33333;
```
I see that other examples have copyright... Well, I'm personally kinda against putting copyright on trivial examples, that may only complicate users' lives - either make them bother to deal with process regarding copyrights, or make them violators. But if there's a policy that every file in the project should have a copyright notice, please re-confirm, and I'll add it.

grgustaf replied:
I guess it would help me, so we don't wrongly slap an Intel copyright on it at some point.

pfalcon replied:
As discussed on IRC, as I'm contributing from my company, to a product of another company, my interpretations are irrelevant, and I'll just follow existing practice of the codebase. Fixed.

pfalcon commented on src/zjs_dgram.c:
```diff
@@ -0,0 +1,334 @@
+                                            const jerry_length_t argc)
+{
+    int ret;
+    if (argc != 1)
```
I "don't know" (ahem) what's "JS common practice" on that, but if it has practice like that, it's a dangerous practice (and JS is a dangerous language to develop programs in). In this case, the check above has very specific goal though: https://nodejs.org/api/dgram.html#dgram_dgram_createsocket_type_callback gives the function prototype:

> dgram.createSocket(type[, callback])

callback is optional and is a shortcut for calling `.on("message", callback)` later. This shortcut is not supported by a small implementation like Zephyr.js, and if user will try to use it, they will be properly warned that there's a problem.

grgustaf replied:
Well, ignoring extra arguments is good enough for C... :)

I see, I didn't realize you were explicitly not supporting that. In that case, it's fine I guess. You might consider returning NOTSUPPORTED_ERROR() if argc >= 2.

pfalcon replied:
> consider returning NOTSUPPORTED_ERROR() if argc >= 2.

Done.

pfalcon commented:
So, as I fix issues raised in review comments, I add new commits, as was requested previously, instead of squashing my changes into existing ones right away.

But the squashing would need to happen eventually, and for the multitude of "fix" commits not sneaking in unsquashed, I purposedly don't sign-off them/use different commit message format.

pfalcon commented:
@grgustaf : So, I believe I addressed all of your comments. Can you please look thru the "fix" commits and let me know if I can squash them back into the main, "content" commits?

And otherwise, on TODO is just adding "error" callback event, as requested by @poussa.

grgustaf commented:
@pfalcon yeah, fixes look fine, you can squash!

pfalcon commented:
Ok, squashed.
